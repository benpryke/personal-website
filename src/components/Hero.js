import React from 'react';

import Banner from './Banner';
import { LinkedInButton, TwitterButton } from './SocialButton';

/**
 * A hero banner
 * 
 * Contains an image and two speech bubbles
 */
export default class Hero extends React.Component {
  static MAX_HEIGHT = 500;
  static MIN_HEIGHT = 100;
  static SCROLL_CUTOFF = Hero.MAX_HEIGHT - Hero.MIN_HEIGHT;

  constructor(props) {
    super(props);

    this.state = {
      height: Hero.MAX_HEIGHT,
    }
    
    this.computeHeight = this.computeHeight.bind(this);
  }

  componentWillMount() {
    this.scrollListener = document.addEventListener('scroll', this.computeHeight);
    this.resizeListener = window.addEventListener('resize', this.computeHeight);
  }

  componentWillUnmount() {
    document.removeEventListener('scroll', this.scrollListener);
    window.removeEventListener('resize', this.resizeListener);
  }

  computeHeight() {
    const top = window.scrollY;
    
    // Match the min-width and min-height media rules for applying a header animation
    if (window.innerWidth < 900 || window.innerHeight < 700) {
      this.setState({
        height: Hero.MAX_HEIGHT,
      });
    } else {
      this.setState({
        height: top < Hero.SCROLL_CUTOFF ? Hero.MAX_HEIGHT - top : Hero.MAX_HEIGHT - Hero.SCROLL_CUTOFF,
      });
    }
  }

  render() {
    const { name, headshot, ...rest } = this.props;
    const headshotSize = this.state.height * 0.8;

    return (
      <Banner className='hero' {...rest}>
        <div className='hero-content' style={{ maxHeight: this.state.height }}>
          <img src={headshot} alt={name} style={{
            width: headshotSize,
            height: headshotSize,
          }}/>
          <div className='conversation'>
            <div className='greeting speech-bubble'>
              <h1>Hi, I'm {name}</h1>
            </div>
            <div className='action speech-bubble'>
              <LinkedInButton/>
              <TwitterButton/>
            </div>
          </div>
        </div>
      </Banner>
    );
  }
}
