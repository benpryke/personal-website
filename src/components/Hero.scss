section.banner.hero {
  position: relative;
  display: flex;
  justify-content: center;
  padding: 130px 10%;
  width: stretch;
  max-width: 100vw;
  top: 0;
  background: $primary-colour;
  overflow: hidden;
  z-index: 1000;
  transition: height 0.3s ease;
  box-sizing: border-box;
}

.hero-content {
  position: relative;
  display: flex;
  flex-direction: column;
  justify-content: space-evenly;
  align-items: center;
  width: inherit;
  height: inherit;
  max-width: $screen-width-l;
  padding: 0 10%;
  animation: hero-fade-in 0.2s ease-out;
}

@keyframes hero-fade-in {
  0% { opacity: 0; }
  100% { opacity: 1; }
}

.hero {
  img {
    position: relative;
    max-width: 80vw;
    max-height: 80vw;
    border-radius: 100%;
    border: 3px solid $primary-colour-light;
    box-shadow: 5px 5px 10px rgba(2, 23, 33, 0.12);
    flex-shrink: 0;
  }

  .greeting {
    z-index: 10;
  }

  .action {
    margin-left: auto;
  }
}

.conversation {
  display: flex;
  flex-direction: column;
  width: 100%;
  margin-top: 20px;
}

@mixin speech-bubble-arrow-base() {
  content: '';
  position: absolute;
  width: 0;
  height: 0;
  border: 20px solid transparent;
}

.speech-bubble {
  position: relative;
  display: flex;
  align-items: center;
  justify-content: center;
  width: 100%;
  max-width: 100%;
  height: 70px;
  margin: 10px 0;
  padding: 0 10px;
  border-radius: 10px;
  background: #fbfeff;
  box-sizing: border-box;
  box-shadow: 5px 5px 10px rgba(2, 23, 33, 0.12);
  
  // Ensure that the speech bubble contents are in front of the bubble arrow
  * {
    z-index: 5;
  }
  
  h1 {
    font-size: 1.2em;
    white-space: nowrap;
  }

  &.greeting:after {
    @include speech-bubble-arrow-base();
    left: 0;
    top: 0;
    border-right-color: #fbfeff;
    border-left: 0;
    border-top: 0;
    margin-top: 0;
    margin-left: -12px;
  }

  &.action:after {
    @include speech-bubble-arrow-base();
    right: 0;
    bottom: 0;
    border-left-color: #fbfeff;
    border-right: 0;
    border-bottom: 0;
    margin-top: 0;
    margin-right: -12px;
  }
}

// After $screen-width-xs, we can increase the header font size
@media only screen and (min-width: $screen-width-xs) {
  .speech-bubble h1 {
    font-size: 1.5em;
  }
}

// Between widths of 900 and 1000 px, the header font size is temporarily too large
@media only screen and (min-width: $screen-width-m) and (max-width: 1000px) {
  .speech-bubble h1 {
    font-size: 1.2em;
  }
}

// Devices with a larger width get a header animation
@media only screen and (min-width: $screen-width-m) {
  section.banner.hero {
    padding: 0;
    height: 500px;
  }

  .hero-content {
    flex-direction: row;
    justify-content: center;
  }

  .hero img {
    width: 400px;
    height: 400px;
  }

  .conversation {
    flex-direction: row;
    flex-wrap: wrap;
    width: auto;
    margin-top: 0;
  }
  
  .speech-bubble {
    width: calc(40vw - 120px);
    max-width: $screen-width-l / 2 - 120px;
    margin: 10px;
  }
}

// Devices with larger width & height get a sticky header
@media only screen and (min-width: $screen-width-m) and (min-height: $screen-height-s) {
  section.banner.hero  {
    position: sticky;
    top: -400px;
  }

  .hero-content {
    position: fixed;
    top: 0;
  }
}

// We need to disable the max-height set from the Hero component on smaller screens
@media only screen and (max-width: $screen-width-m - 0.0001px) {
  .hero-content {
    max-height: none !important;
  }
}
