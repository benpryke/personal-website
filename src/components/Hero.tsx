import React from "react";

import Banner, { BannerPassthroughProps } from "./Banner";
import { LinkedInButton, TwitterButton } from "./SocialButton";

export interface HeroProps extends BannerPassthroughProps {
  /** Name of our protagonist */
  name: string;
  /** URL to a headshot of the protagonist */
  headshot: string;
}

interface HeroState {
  height: number;
}

/**
 * A hero banner
 *
 * Contains an image and two speech bubbles
 */
export default class Hero extends React.Component<HeroProps, HeroState> {
  static MAX_HEIGHT = 500;
  static MIN_HEIGHT = 100;
  static SCROLL_CUTOFF = Hero.MAX_HEIGHT - Hero.MIN_HEIGHT;
  static MIN_ANIM_WINDOW_WIDTH = 900;
  static MIN_ANIM_WINDOW_HEIGHT = 700;

  readonly state: HeroState = {
    height: Hero.computeHeight(),
  };

  constructor(props: HeroProps) {
    super(props);
    this.setHeight = this.setHeight.bind(this);
  }

  componentDidMount() {
    document.addEventListener("scroll", this.setHeight);
    window.addEventListener("resize", this.setHeight);

    // As we use react-snap and ReactDOM.hydrate in prod, if the user refreshes
    // the page while scrolled down the .hero-content max-height will mismatch.
    // This effectively freezes us in a broken state unless we update height.
    // We can fix this with the following double-render hack.
    if (window.scrollY > 0) {
      this.setState({ height: this.state.height + 0.0001 });
    }
  }

  componentWillUnmount() {
    document.removeEventListener("scroll", this.setHeight);
    window.removeEventListener("resize", this.setHeight);
  }

  static computeHeight(): number {
    // Match the min-width and min-height media rules for applying a header animation
    if (
      window.innerWidth < Hero.MIN_ANIM_WINDOW_WIDTH ||
      window.innerHeight < Hero.MIN_ANIM_WINDOW_HEIGHT
    ) {
      console.log("1", Hero.MAX_HEIGHT);
      return Hero.MAX_HEIGHT;
    } else {
      const top = window.scrollY;
      console.log(
        "2",
        top,
        top < Hero.SCROLL_CUTOFF,
        top < Hero.SCROLL_CUTOFF
          ? Hero.MAX_HEIGHT - top
          : Hero.MAX_HEIGHT - Hero.SCROLL_CUTOFF
      );
      return top < Hero.SCROLL_CUTOFF
        ? Hero.MAX_HEIGHT - top
        : Hero.MAX_HEIGHT - Hero.SCROLL_CUTOFF;
    }
  }

  setHeight(): void {
    this.setState({ height: Hero.computeHeight() });
  }

  render() {
    const { name, headshot, ...rest } = this.props;
    const { height } = this.state;
    const headshotSize = height * 0.8;

    return (
      <Banner className="hero" {...rest}>
        <div className="hero-content" style={{ maxHeight: Math.floor(height) }}>
          <img
            src={headshot}
            alt={name}
            style={{
              width: headshotSize,
              height: headshotSize,
            }}
          />
          <div className="conversation">
            <div className="greeting speech-bubble">
              <h1>Hi, I'm {name}</h1>
            </div>
            <div className="action speech-bubble">
              <LinkedInButton />
              <TwitterButton />
            </div>
          </div>
        </div>
      </Banner>
    );
  }
}
