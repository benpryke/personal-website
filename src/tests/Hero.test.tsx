import React from 'react';
import { mount, shallow } from 'enzyme';
import renderer from 'react-test-renderer';

import Hero, { HeroProps } from '../components/Hero';
import { setWindowSize, setScrollY } from './utils';

const props: HeroProps = {
  name: 'name',
  headshot: 'url',
  fadeIn: false,
  fadeInOffset: 200,
  className: 'cls',
  style: {color: 'red'},
};

describe('Hero', () => {
  it('renders without crashing', () => {
    mount(<Hero {...props}/>);
  });

  it('should match the snapshot', () => {
    const tree = renderer.create(<Hero {...props}/>).toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('renders with a fixed height when the window is small', () => {
    const wrapper = mount<Hero>(<Hero {...props}/>);
    setScrollY(Hero.MAX_HEIGHT * 2);
    
    // Small width
    setWindowSize(Hero.MIN_ANIM_WINDOW_WIDTH - 1, Hero.MIN_ANIM_WINDOW_HEIGHT);
    expect(wrapper.state().height).toEqual(Hero.MAX_HEIGHT);
    
    // Small height
    setWindowSize(Hero.MIN_ANIM_WINDOW_WIDTH, Hero.MIN_ANIM_WINDOW_HEIGHT - 1);
    expect(wrapper.state().height).toEqual(Hero.MAX_HEIGHT);
  });
  
  it('renders with a variable height when the window is large', () => {
    const wrapper = mount<Hero>(<Hero {...props}/>);
    setWindowSize(Hero.MIN_ANIM_WINDOW_WIDTH, Hero.MIN_ANIM_WINDOW_HEIGHT);
    
    // Scrolled to top
    setScrollY(0);
    expect(wrapper.state().height).toEqual(Hero.MAX_HEIGHT);
    
    // Scrolled to bottom
    setScrollY(Hero.MAX_HEIGHT * 2);
    expect(wrapper.state().height).toEqual(Hero.MIN_HEIGHT);
  });
});
