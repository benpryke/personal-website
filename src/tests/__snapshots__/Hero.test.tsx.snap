// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Hero should match the snapshot 1`] = `
<section
  className="banner cls"
  style={
    Object {
      "color": "red",
    }
  }
>
  <div
    className="hero-content"
    style={
      Object {
        "maxHeight": 500,
      }
    }
  >
    <img
      alt="name"
      src="url"
      style={
        Object {
          "height": 400,
          "width": 400,
        }
      }
    />
    <div
      className="conversation"
    >
      <div
        className="greeting speech-bubble"
      >
        <h1>
          Hi, I'm 
          name
        </h1>
      </div>
      <div
        className="action speech-bubble"
      >
        <a
          className="social-button"
          href="https://linkedin.com/in/benjaminpryke"
        >
          <i
            className="fab fa-linkedin"
          />
          LinkedIn
        </a>
        <a
          className="social-button"
          href="https://twitter.com/BenjaminPryke"
        >
          <i
            className="fab fa-twitter"
          />
          Twitter
        </a>
      </div>
    </div>
  </div>
</section>
`;
